src/composition.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024 the Interpoli Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use alloc::{string::String, vec::Vec};
+use core::ops::Range;
+use kurbo::{PathEl, Shape as _};
+
+#[cfg(feature = "std")]
+use std::collections::HashMap;
+
+use crate::{animated, Brush, Repeater, Stroke, Transform, Value};
+
+/// Model of a Lottie file.
+#[derive(Clone, Default, Debug)]
+pub struct Composition {
+    /// Frames in which the animation is active.
+    pub frames: Range<f64>,
+    /// Frames per second.
+    pub frame_rate: f64,
+    /// Width of the animation.
+    pub width: usize,
+    /// Height of the animation.
+    pub height: usize,
+    /// Precomposed layers that may be instanced.
+    #[cfg(feature = "std")]
+    pub assets: HashMap<String, Vec<Layer>>,
+    /// Collection of layers.
+    pub layers: Vec<Layer>,
+}
+
+#[derive(Clone, Debug)]
+pub enum Geometry {
+    Fixed(Vec<PathEl>),
+    Rect(animated::Rect),
+    Ellipse(animated::Ellipse),
+    Spline(animated::Spline),
+}
+
+impl Geometry {
+    pub fn evaluate(&self, frame: f64, path: &mut Vec<PathEl>) {
+        match self {
+            Self::Fixed(value) => {
+                path.extend_from_slice(value);
+            }
+            Self::Rect(value) => {
+                path.extend(value.evaluate(frame).path_elements(0.1));
+            }
+            Self::Ellipse(value) => {
+                path.extend(value.evaluate(frame).path_elements(0.1));
+            }
+            Self::Spline(value) => {
+                value.evaluate(frame, path);
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Draw {
+    /// Parameters for a stroked draw operation.
+    pub stroke: Option<Stroke>,
+    /// Brush for the draw operation.
+    pub brush: Brush,
+    /// Opacity of the draw operation.
+    pub opacity: Value<f64>,
+}
+
+/// Elements of a shape layer.
+#[derive(Clone, Debug)]
+pub enum Shape {
+    /// Group of shapes with an optional transform.
+    Group(Vec<Shape>, Option<GroupTransform>),
+    /// Geometry element.
+    Geometry(Geometry),
+    /// Fill or stroke element.
+    Draw(Draw),
+    /// Repeater element.
+    Repeater(Repeater),
+}
+
+/// Transform and opacity for a shape group.
+#[derive(Clone, Debug)]
+pub struct GroupTransform {
+    pub transform: Transform,
+    pub opacity: Value<f64>,
+}
+
+/// Layer in an animation.
+#[derive(Clone, Debug, Default)]
+pub struct Layer {
+    /// Name of the layer.
+    pub name: String,
+    /// Index of the transform parent layer.
+    pub parent: Option<usize>,
+    /// Transform for the entire layer.
+    pub transform: Transform,
+    /// Opacity for the entire layer.
+    pub opacity: Value<f64>,
+    /// Width of the layer.
+    pub width: f64,
+    /// Height of the layer.
+    pub height: f64,
+    /// Blend mode for the layer.
+    pub blend_mode: Option<peniko::BlendMode>,
+    /// Range of frames in which the layer is active.
```

Comment:
These are hax to address the design issue of putting `frame_rate` on `Composition`, which should be addressed by the change in #17

Reply:
`frames` is not related to that ... That's so that a layer can be active for some subset of the full composition.

Reply:
They are relative to the `Composition`'s rate though, right?